docs/source/extending_pywr/extending_pywr_nodes.rst

```diff
@@ -0,0 +1,116 @@
+(``IN_1``) which conveys water downstream, connected by a set of links in parallel (``LNK_1`` ... ``LNK_N``) each with
+a different ``max_flow`` and ``cost``, illustrated below::
+
+                               /-->-- LNK_1 -->--\
+    UPSTREAM -->-- OUT_1 -->--|--->--  ...  -->---|-->-- IN_1 -->-- DOWNSTREAM
+                               \-->-- LNK_N -->--/
+
+Let's look at an example to create a new node type that represents a leaky pipe. To remove water from the system we
+need to use an output node (``LEAK``), with two links representing the boundaries of the compound node (``INFLOW`` and
+``OUTFLOW``)::
+
+    UPSTREAM -->-- INFLOW -->-- OUTFLOW -->-- DOWNSTREAM
+                     |
+                     \------>--  LEAK
+
+This is a simple structure which represents leakage as a demand with a maximum value and a benefit to be supplied. It
+is slightly flawed as the leakage volume does not vary proportionally to flow through the link, but is sufficient as
+an example:
+
+.. code-block:: python
+
+    from pywr.nodes import Node, Link, Output
+
+    class LeakyPipe(Node):
+        def __init__(self, leakage, leakage_cost=-99999, *args, **kwargs):
+            self.allow_isolated = True  # Required for compound nodes
+
+            super(LeakyPipe, self).__init__(*args, **kwargs)
+
+            # Define the internal nodes
+            self.inflow = Link(self.model, name='{} In'.format(self.name), parent=self)
+            self.outflow = Link(self.model, name='{} Out'.format(self.name), parent=self)
+            self.leak = Output(self.model, name='{} Leak'.format(self.name), parent=self)
+
+            # Connect the internal nodes
+            self.inflow.connect(self.outflow)
+            self.inflow.connect(self.leak)
+
+            # Define the properties of the leak (demand and benefit)
+            self.leak.max_flow = leakage
+            self.leak.cost = leakage_cost
+
+        def iter_slots(self, slot_name=None, is_connector=True):
+            # This is required so that connecting to this node actually connects to the outflow sub-node, and
+            # connecting from this node actually connects to the input sub-node
+            if is_connector:
+                yield self.outflow
+            else:
+                yield self.inflow
+
+        def after(self, timestep):
+            # Make the flow on the compound node appear as the flow _after_ the leak
+            self.commit_all(self.outflow.flow)
+            # Make sure save is done after setting aggregated flow
+            super(LeakyPipe, self).after(timestep)
+
+        @classmethod
+        def load(cls, data, model):
+            del(data["type"])
+            leakage = data.pop("leakage")
+            leakage_cost = data.pop("leakage_cost", None)
+            return cls(model, leakage, leakage_cost, **data)
+
+
+The custom node does not need to be "registered" manually (unlike parameters, which do). They will work in JSON as
```

Comment:
Worth mentioning metaclasses here for the advanced reader?

Reply:
I did consider it. I can't bring myself to say it without the word "magic" too!

Reply:
Are we not allowed to use the phrase "through the magic of metaclasses!"?